### PR TITLE
NMS-12676: Basic Service tracker and integration

### PR DESCRIPTION
**Pull request by fooker**
* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12676

### CODE REVIEW

File: opennms-dao-api/src/main/java/org/opennms/netmgt/dao/api/ApplicationDao.java [hunk 1]
```diff
@@ -70,4 +73,6 @@
     List<MonitoredServiceStatusEntity> getAlarmStatus();
 
     List<MonitoredServiceStatusEntity> getAlarmStatus(List<OnmsApplication> applications);
+
+    List<OnmsMonitoringLocation> getPerspeciveLocationsForService(final int nodeId, final InetAddress ipAddress, final String serviceName);
```
Comment (christianpape) [inline opennms-dao-api/src/main/java/org/opennms/netmgt/dao/api/ApplicationDao.java:77]: typo, Perspecive -> Perspective

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/ApplicationDaoHibernate.java [hunk 2]
```diff
@@ -145,4 +147,15 @@ public OnmsApplication findByName(final String name) {
 		}
 		return entityList;
 	}
+
+	public List<OnmsMonitoringLocation> getPerspeciveLocationsForService(final int nodeId, final InetAddress ipAddress, final String serviceName) {
+		return (List<OnmsMonitoringLocation>) getHibernateTemplate().find("select distinct perspeciveLocation " +
+																		  "from OnmsMonitoredService service " +
+																		  "join service.applications application " +
+																		  "join application.perspectiveLocations perspeciveLocation " +
```
Comment (christianpape) [inline opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/ApplicationDaoHibernate.java:155]: typo

**Comment by fooker** [comment]
Basic structure is done, I think.
I would like to add some more tests for the service tracker.

File: features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/RemotePollerd.java [hunk 3]
```diff
@@ -118,336 +117,253 @@
...
-            return null;
+            try {
+                this.scheduler.scheduleJob(job, trigger);
+            } catch (final SchedulerException e) {
+                LOG.error("Failed to schedule {} ({}).", remotePolledService, key, e);
+            }
         });
     }
 
+    private void delService(final ServiceTracker.ServiceEntry<Set<RemotePolledService>> entry) {
```
Comment (patrick-schweizer) [inline features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/RemotePollerd.java:277]: why not deleteService()?

File: features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/RemotePollerd.java [hunk 4]
```diff
@@ -483,16 +399,20 @@ public void saveResponseTimeData(final String locationName, final RemotePolledSe
...
+        try {
+            polledService.getThresholdingSession().accept(collectionSetDTO);
+        } catch (final Throwable e) {
+            LOG.error("Failed to threshold on {} for {} because of an exception", polledService, dsName, e);
+        }
     }
 
-    private String getServiceParameter(final Service svc, final String key) {
-        for(final Parameter parm : pollerConfig.parameters(svc)) {
+    private String getServiceParameter(final Service service, final String key) {
```
Comment (patrick-schweizer) [inline features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/RemotePollerd.java:414]: +1 on using readable variables - makes the code much easier to read...

File: features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/ServiceTracker.java [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+        entry.ifPresent(this.delService);
+
+    }
+
+    @EventHandler(ueis = {EventConstants.NODE_DELETED_EVENT_UEI,
+                          EventConstants.DUP_NODE_DELETED_EVENT_UEI})
+    public void nodeDeletedHandler(final IEvent event) {
+        final Node node = Node.fromEvent(event);
+
+//        this.getQueryManager().closeOutagesForNode(event.getTime(), event.getDbid(), node.nodeId);
```
Comment (patrick-schweizer) [inline features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/ServiceTracker.java:223]: there are quite some lines of commented code in this class. Not sure if you want to remove right away or do you still need it for a later refactoring?

File: features/remotepollerng/daemon/src/test/java/org/opennms/netmgt/remotepollerng/RemotePollerdIT.java [hunk 6]
```diff
@@ -205,92 +211,105 @@ public void afterTransaction() throws Exception {
...
+                                                 .addParam(EventConstants.PARM_DAEMON_NAME, RemotePollerd.NAME)
+                                                 .getEvent());
     }
 
-    private void sendPollingPackageAssociationChanged(final String locationName) {
-        final EventBuilder ebldr = new EventBuilder(EventConstants.POLLER_PACKAGE_LOCATION_ASSOCIATION_CHANGED_EVENT_UEI, "test");
-        ebldr.addParam(EventConstants.PARM_LOCATION, locationName);
-        this.eventIpcManager.sendNow(ebldr.getEvent());
-    }
+//    private void changePollingPackages(final String locationName, final String ... packages) {
```
Comment (patrick-schweizer) [inline features/remotepollerng/daemon/src/test/java/org/opennms/netmgt/remotepollerng/RemotePollerdIT.java:221]: remove?

File: features/remotepollerng/daemon/src/test/java/org/opennms/netmgt/remotepollerng/ServiceTrackerIT.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        "classpath:/META-INF/opennms/applicationContext-testThresholdingDaos.xml",
+        "classpath:/META-INF/opennms/applicationContext-testPollerConfigDaos.xml",
+        "classpath:/META-INF/opennms/applicationContext-rpc-utils.xml",
+})
+@JUnitConfigurationEnvironment
+@JUnitTemporaryDatabase(reuseDatabase = false)
+@Ignore
+public class ServiceTrackerIT {
+
+//    @Autowired
```
Comment (patrick-schweizer) [inline features/remotepollerng/daemon/src/test/java/org/opennms/netmgt/remotepollerng/ServiceTrackerIT.java:84]: add a TODO to clean up the test later?

File: opennms-base-assembly/src/main/filtered/etc/events/opennms.remote.poller.events.xml [hunk 8]
```diff
@@ -30,6 +30,7 @@
     </event>
 
     <event>
+        <!-- TODO: Renmove me -->
```
Comment (patrick-schweizer) [inline opennms-base-assembly/src/main/filtered/etc/events/opennms.remote.poller.events.xml:33]: add your to the TODO as to easier filter whose TODO it is?

**Comment by patrick-schweizer** [comment]
I added a few minor comments.

File: features/remotepollerng/daemon/src/test/java/org/opennms/netmgt/remotepollerng/RemotePollerdIT.java [hunk 9]
```diff
@@ -97,7 +98,6 @@
 })
 @JUnitConfigurationEnvironment
 @JUnitTemporaryDatabase(reuseDatabase = false)
-@Ignore // TODO: Patrick: rewrite test after we have finalized how it should look like...
 public class RemotePollerdIT implements InitializingBean {
```
Comment (christianpape) [inline features/remotepollerng/daemon/src/test/java/org/opennms/netmgt/remotepollerng/RemotePollerdIT.java:101]: Nice. That was a lot of work.

**Review by christianpape — APPROVED** [review]

File: features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/ServiceTracker.java [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+
+        public Service service(final String serviceName) {
+            return new Service(this.nodeId, this.ipAddress, serviceName);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                              .add("nodeId", this.nodeId)
+                              .add("address", this.ipAddress)
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/ServiceTracker.java:398]: :+1:

File: features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/ServiceTracker.java [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+        @Override
+        public int hashCode() {
+            return Objects.hash(this.nodeId, this.ipAddress, this.serviceName);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                              .add("nodeId", this.nodeId)
+                              .add("address", this.ipAddress)
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/ServiceTracker.java:456]: :+1:

**Pull request merged**